bigquery_etl/backfill/shredder_mitigation.py

```diff
@@ -50,30 +67,187 @@ class DataTypeGroup(Enum):
     UNDETERMINED = "None"
 
 
+@attr.define
 class Column:
     """Representation of a column in a query, with relevant details for shredder mitigation."""
 
-    def __init__(self, name, data_type, column_type, status):
-        """Initialize class with required attributes."""
-        self.name = name
-        self.data_type = data_type
-        self.column_type = column_type
-        self.status = status
-
-    def __eq__(self, other):
-        """Return attributes only if the referenced object is of type Column."""
-        if not isinstance(other, Column):
-            return NotImplemented
-        return (
-            self.name == other.name
-            and self.data_type == other.data_type
-            and self.column_type == other.column_type
-            and self.status == other.status
+    name: str
+    data_type: DataTypeGroup = attr.field(default=DataTypeGroup.UNDETERMINED)
+    column_type: ColumnType = attr.field(default=ColumnType.UNDETERMINED)
+    status: ColumnStatus = attr.field(default=ColumnStatus.UNDETERMINED)
+
+    """Validate the type of the attributes."""
+
+    @data_type.validator
+    def validate_data_type(self, attribute, value):
+        """Check that the type of data_type is as expected."""
+        if not isinstance(value, DataTypeGroup):
+            raise ValueError(f"Invalid {value} with type: {type(value)}.")
+
+    @column_type.validator
+    def validate_column_type(self, attribute, value):
+        """Check that the type of parameter column_type is as expected."""
+        if not isinstance(value, ColumnType):
+            raise ValueError(f"Invalid data type for: {value}.")
+
+    @status.validator
+    def validate_status(self, attribute, value):
+        """Check that the type of parameter column_status is as expected."""
+        if not isinstance(value, ColumnStatus):
+            raise ValueError(f"Invalid data type for: {value}.")
+
+
+@attr.define
+class Subset:
+    client: bigquery.Client()
+    destination_table: str = attr.field()
```

Comment:
I'm slightly confused by the parameters

is the `destination_table` with or without the `_v3`?

From the version function I assume with ... but what is the suffix then? in the `name()` function you add the suffix to it and I'm not sure what that would be... is it related to partitioning?

Reply:
Yes, the destination_table includes the version.
The intention of the suffix was to enable saving the result of the subset results to temp tables which would have e.g. timestamp as suffix. However the subsets are calculated in a single query which makes suffixes an unnecessary complication, thus removed. Please do unresolve if needed!